File: apps/uploader/src/uploader/domain/models/radiology.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ImagingSeriesBase:
    source_id: str
    series_identifier: str | None
    imaging_study_identifier: str | None
    dicom_images_count: int | None = None
    series_start_date: str | None = None
    body_region: str | None = None
    laterality: str | None = None
    imaging_device: str | None = None
    manufacturer_of_imaging_device: str | None = None
    software_version: str | None = None
    color_space: str | None = None
    pixel_spacing: int | None = None
    image_type: str | None = None
    file_format: str | None = None
    file_size: int | None = None
    image_width: int | None = None
    image_height: int | None = None
    image_depth: int | None = None
    number_of_channels: int | None = None
    channel_resolution: int | None = None
    compression_method: str | None = None
    compression_quality_label: str | None = None
    annotations_available: bool | None = None


@dataclass(frozen=True)
class CtSeries(ImagingSeriesBase):
    tube_voltage_kvp: int | None = None
    x_ray_tube_current_ma: int | None = None
    exposure_time_ms: int | None = None
    spiral_pitch_factor: float | None = None
    filter_type: str | None = None
    convolution_kernel: str | None = None
    field_of_view: float | None = None
    slice_thickness: float | None = None
    imaging_injection: str | None = None
    number_of_image_planes: int | None = None


@dataclass(frozen=True)
class MrSeries(ImagingSeriesBase):
    sequence_name: str | None = None
    magnetic_field_strength: float | None = None
    mr_acquisition_type: str | None = None
    repetition_time: float | None = None
    echo_time: float | None = None
    imaging_frequency: float | None = None
    flip_angle: int | None = None
    inversion_time: int | None = None
    receive_coil_name: str | None = None
    field_of_view: float | None = None
    slice_thickness: float | None = None
    imaging_injection: str | None = None
    number_of_image_planes: int | None = None


@dataclass(frozen=True)
class DxSeries(ImagingSeriesBase):
    patient_orientation: str | None = None
    tube_voltage_kvp: int | None = None
    exposure_time_ms: int | None = None
    exposure_mas: int | None = None


@dataclass(frozen=True)
class MgSeries(ImagingSeriesBase):
    tube_voltage_kvp: int | None = None
    exposure_time_ms: int | None = None
    exposure_mas: int | None = None
    compression_force_n: float | None = None


@dataclass(frozen=True)
class UsSeries(ImagingSeriesBase):
    transducer_frequency_mhz: float | None = None
    mechanical_index: float | None = None
    thermal_index: float | None = None


@dataclass(frozen=True)
class ImagingStudy:
    accession_number: str
    source_id: str
    imaging_study_identifier: str | None = None
    belongs_to_person: str | None = None
    imaging_modality: list[str] | None = None
    body_region: list[str] | None = None
    imaging_procedure: list[str] | None = None
    reason_for_imaging_procedure: list[str] | None = None
    study_start_date: str | None = None
    dicom_series_count: int | None = None
    dicom_images_count: int | None = None
    affiliated_institution: str | None = None
    ct_series: CtSeries | None = None
    mr_series: MrSeries | None = None
    us_series: UsSeries | None = None
    dx_series: DxSeries | None = None
    mg_series: MgSeries | None = None


RadiologyData = list[ImagingStudy]

File: src/Uploader/Uploader.Domain/Radiology.cs
using Uploader.Domain.Common;

namespace Uploader.Domain;

/// <summary>Shared DICOM imaging-series attributes (report/FAIR Genomes).</summary>
public abstract record ImagingSeriesBase : Entity
{
    public required string SourceId { get; init; }
    public string? SeriesIdentifier { get; init; }
    public string? ImagingStudyIdentifier { get; init; }
    public int? DicomImagesCount { get; init; }
    public string? SeriesStartDate { get; init; }
    public string? BodyRegion { get; init; }
    public string? Laterality { get; init; }
    public string? ImagingDevice { get; init; }
    public string? ManufacturerOfImagingDevice { get; init; }
    public string? SoftwareVersion { get; init; }
    public string? ColorSpace { get; init; }
    public int? PixelSpacing { get; init; }
    public string? ImageType { get; init; }
    public string? FileFormat { get; init; }
    public int? FileSize { get; init; }
    public int? ImageWidth { get; init; }
    public int? ImageHeight { get; init; }
    public int? ImageDepth { get; init; }
    public int? NumberOfChannels { get; init; }
    public int? ChannelResolution { get; init; }
    public string? CompressionMethod { get; init; }
    public string? CompressionQualityLabel { get; init; }
    public bool? AnnotationsAvailable { get; init; }
}

/// <summary>CT imaging series.</summary>
public sealed record CtSeries : ImagingSeriesBase
{
    public int? TubeVoltageKvp { get; init; }
    public int? XRayTubeCurrentMa { get; init; }
    public int? ExposureTimeMs { get; init; }
    public double? SpiralPitchFactor { get; init; }
    public string? FilterType { get; init; }
    public string? ConvolutionKernel { get; init; }
    public double? FieldOfView { get; init; }
    public double? SliceThickness { get; init; }
    public string? ImagingInjection { get; init; }
    public int? NumberOfImagePlanes { get; init; }
}

/// <summary>MR imaging series.</summary>
public sealed record MrSeries : ImagingSeriesBase
{
    public string? SequenceName { get; init; }
    public double? MagneticFieldStrength { get; init; }
    public string? MrAcquisitionType { get; init; }
    public double? RepetitionTime { get; init; }
    public double? EchoTime { get; init; }
    public double? ImagingFrequency { get; init; }
    public int? FlipAngle { get; init; }
    public int? InversionTime { get; init; }
    public string? ReceiveCoilName { get; init; }
    public double? FieldOfView { get; init; }
    public double? SliceThickness { get; init; }
    public string? ImagingInjection { get; init; }
    public int? NumberOfImagePlanes { get; init; }
}

/// <summary>DX (digital radiography) imaging series.</summary>
public sealed record DxSeries : ImagingSeriesBase
{
    public string? PatientOrientation { get; init; }
    public int? TubeVoltageKvp { get; init; }
    public int? ExposureTimeMs { get; init; }
    public int? ExposureMas { get; init; }
}

/// <summary>MG (mammography) imaging series.</summary>
public sealed record MgSeries : ImagingSeriesBase
{
    public int? TubeVoltageKvp { get; init; }
    public int? ExposureTimeMs { get; init; }
    public int? ExposureMas { get; init; }
    public double? CompressionForceN { get; init; }
}

/// <summary>US (ultrasound) imaging series.</summary>
public sealed record UsSeries : ImagingSeriesBase
{
    public double? TransducerFrequencyMhz { get; init; }
    public double? MechanicalIndex { get; init; }
    public double? ThermalIndex { get; init; }
}

/// <summary>A radiology imaging study keyed by accession number.</summary>
public sealed record ImagingStudy : Entity
{
    public required string AccessionNumber { get; init; }
    public required string SourceId { get; init; }
    public string? ImagingStudyIdentifier { get; init; }
    public string? BelongsToPerson { get; init; }
    public IReadOnlyList<string>? ImagingModality { get; init; }
    public IReadOnlyList<string>? BodyRegion { get; init; }
    public IReadOnlyList<string>? ImagingProcedure { get; init; }
    public IReadOnlyList<string>? ReasonForImagingProcedure { get; init; }
    public string? StudyStartDate { get; init; }
    public int? DicomSeriesCount { get; init; }
    public int? DicomImagesCount { get; init; }
    public string? AffiliatedInstitution { get; init; }
    public CtSeries? CtSeries { get; init; }
    public MrSeries? MrSeries { get; init; }
    public UsSeries? UsSeries { get; init; }
    public DxSeries? DxSeries { get; init; }
    public MgSeries? MgSeries { get; init; }
}
